libgdf/src/gdf_type_dispatcher.cuh

```diff
@@ -0,0 +1,60 @@
+#ifndef GDF_TYPE_DISPATCHER_H
+#define GDF_TYPE_DISPATCHER_H
+
+#include <utility>
+#include <gdf/cffi/types.h>
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief  Invokes a templated functor by dispatching the appropriate type from 
+ * a gdf_dtype enum value.
+ *
+ * This helper function accepts any *functor* (generic lambdas will also work, but
+ * nvcc does not support generic extended lambdas) with an "operator()" template. 
+ * The template may have 1 or more template parameters, but the first parameter must 
+ * be the type dispatched from the  gdf_dtype enum. The remaining template arguments 
+ * must be able to be automatically deduced. 
+ * 
+ * @Param dtype The gdf_dtype enum that determines which type will be dispatched
+ * @Param f The functor with a templated "operator()" that will be invoked with 
+ * the dispatched type
+ * @Param args A parameter-pack (i.e., arbitrary number of arguments) that will 
+ * be perfectly-forwarded as the arguments of the functor's "operator()".
+ *
+ * @Returns Whatever is returned by the functor's "operator()". 
+ *
+ * NOTE: The return type for all template instantiations of the functor's "operator()" 
+ * must be the same, otherwise there will be a compiler error as you would be 
+ * trying to return different types from the same function.
+ */
+/* ----------------------------------------------------------------------------*/
+// This pragma disables a compiler warning that complains about the valid usage
+// of calling a __host__ functor from this function which is __host__ __device__
+#pragma hd_warning_disable
```

Comment:
Ugh. How is it valid usage? Shouldn't you just make the functors __host__ __device__?

Reply:
Imagine I had a functor that had host only code, e.g., a `cudaMalloc` call. Then it wouldn't compile if I made it `__host__ __device__` because the device implementation wouldn't be valid.

Reply:
OK... so technically the really correct way to do this is have separate host and device dispatchers. I'm fine with not doing that though, since I don't think it's likely NVCC will change this to an error soon.